internal/validators/rbac_permissions.go

```diff
@@ -12,101 +13,340 @@ const (
 		}
+		permission := permissions[0]
+		if permission.Actions == nil {
+			return result{}, errNil("deny assignment Actions")
+		}
+		actions := []string{}
+		for _, ptr := range permission.Actions {
+			if ptr == nil {
+				return result{}, errNil("deny assignment Action")
+			}
+			action := *ptr
+			if numWildcards(action) > 1 {
+				return result{}, fmt.Errorf("deny assignment Action %s has multiple wildcards", action)
+			}
+			appendStr(&actions, action)
+		}
+		if permission.NotActions == nil {
+			return result{}, errNil("deny assignment NotActions")
+		}
+		notActions := []string{}
+		for _, ptr := range permission.NotActions {
+			if ptr == nil {
+				return result{}, errNil("deny assignment NotAction")
+			}
+			notAction := *ptr
+			if numWildcards(notAction) > 1 {
+				return result{}, fmt.Errorf("deny assignment NotAction %s has multiple wildcards", notAction)
+			}
+			appendStr(&notActions, notAction)
+		}
+		appendDenyInfo(&denyAssignmentInfoControl, denyAssignmentInfo{
+			actions:    actions,
+			notActions: notActions,
+			id:         denyAssignmentID,
+		})
+		if permission.DataActions == nil {
+			return result{}, errNil("deny assignment DataActions")
+		}
+		dataActions := []string{}
+		for _, ptr := range permission.DataActions {
+			if ptr == nil {
+				return result{}, errNil("deny assignment DataAction")
+			}
+			dataAction := *ptr
+			if numWildcards(dataAction) > 1 {
+				return result{}, fmt.Errorf("deny assignment DataAction %s has multiple wildcards", dataAction)
+			}
+			appendStr(&dataActions, dataAction)
+		}
+		if permission.NotDataActions == nil {
+			return result{}, errNil("deny assignment NotDataActions")
+		}
+		notDataActions := []string{}
+		for _, ptr := range permission.NotDataActions {
+			if ptr == nil {
+				return result{}, errNil("deny assignment NotDataAction")
+			}
+			notDataAction := *ptr
+			if numWildcards(notDataAction) > 1 {
+				return result{}, fmt.Errorf("deny assignment NotDataAction %s has multiple wildcards", notDataAction)
+			}
+			appendStr(&notDataActions, notDataAction)
+		}
+		appendDenyInfo(&denyAssignmentInfoData, denyAssignmentInfo{
+			actions:    dataActions,
+			notActions: notDataActions,
+			id:         denyAssignmentID,
+		})
 	}
-	for _, na := range notActions {
-		if len(na) == 0 {
-			return result{}, fmt.Errorf("invalid NotAction or NotDataAction in current role data, is empty string")
+	for _, role := range roles {
+		if role == nil {
+			return result{}, errNil("role")
+		}
+		if role.Properties == nil {
+			return result{}, errNil("role properties")
 		}
-		if numWildcards(na) > 1 {
-			return result{}, fmt.Errorf("invalid NotAction or NotDataAction in current role data, has multiple wildcards")
+		if role.Properties.Permissions == nil {
+			return result{}, errNil("role properties permissions")
 		}
+		permissions := role.Properties.Permissions
+		// We can expect there to only be one permissions item in the list of permissions. That's
```

Comment:
:wrench: Lots of duplication here. Please handle the dereferencing for denyAssignment permissions & roles in a common helper.

Reply:
I thought of doing this, with `Permissions` from the Azure lib being the common data structure. But it turns out they use two different structs for deny assignments vs. roles. I thought I could do it still with interfaces, but that would end up having all the code anyways, in methods of custom structs I'd make instead of just in this function here. Any thoughts on how I could make a helper and end up with less code?

Reply:
Ah. I didn't look closely enough to realize that they're different structs. That's unfortunate. You could solve it with generics, but that can be a refactor for later.